<h1 style="background-color: #008000; font-family: newtimeroman; font-size: 300%; color: white; border-radius: 50px 20px; padding: 25px; text-align: center;">
  <b> CREDIT SCORE EDA </b>
</h1>

A credit score is often regarded as a financial fingerprint, providing a concise measure of one’s trustworthiness with borrowed money. For a global finance company, the challenge of assessing credit risk is a priority, and data science offers an intelligent solution. Over the years, the company has collected extensive customer data, including demographic details, financial histories, and payment patterns, aiming to classify individuals into distinct credit score brackets and reduce manual effort. This dataset provides a comprehensive view of these factors, with 28 features ranging from annual income and outstanding debt to credit utilization ratios and payment behaviors, all of which play a vital role in evaluating credit risk. By cleaning and analyzing this data, the goal is to identify the most impactful features for training Machine Learning and Deep Learning models, enabling lenders to make swift, accurate decisions. Exploring this data not only advances financial intelligence but also reveals fascinating insights into the elements influencing credit scores.

<div style="color:#009de1">
    
<span style="font-size:22px;font-family:Tahoma;color:#001de1"> About the Dataset </span>
<hr style="height:0.6px;background-color:#009de1">
</div>

<p style="font-family:tahoma;text-align:justify" >


- **Dataset:** Credit Score Dataset   
- **Number of Rows:** 100000
- **Number of Columns:** 28  

| **No** | **INPUTS**                  | **Description**                                                                                  |
|-------|-----------------------------|--------------------------------------------------------------------------------------------------|
| 1     | **ID**                       | Unique identifier for each record.                                                               |
| 2     | **Customer_ID**              | Unique identifier for each customer.                                                             |
| 3     | **Month**                    | Month of the transaction or record.                                                              |
| 4     | **Name**                     | Customer’s name.                                                                                 |
| 5     | **Age**                      | The customer’s age.                                                                              |
| 6     | **SSN**                      | Customer’s social security number.                                                               |
| 7     | **Occupation**               | The customer’s occupation.                                                                       |
| 8     | **Annual_Income**            | The customer’s annual income.                                                                    |
| 9     | **Monthly_Inhand_Salary**    | The customer’s monthly take-home salary.                                                         |
| 10    | **Num_Bank_Accounts**        | Total number of bank accounts owned by the customer.                                            |
| 11    | **Num_Credit_Card**          | Total number of credit cards held by the customer.                                              |
| 12    | **Interest_Rate**            | The interest rate applied to loans or credits.                                                  |
| 13    | **Num_of_Loan**              | Number of loans the customer has taken.                                                          |
| 14    | **Type_of_Loan**             | Categories of loans obtained by the customer.                                                   |
| 15    | **Delay_from_due_date**      | The delay in payment relative to the due date.                                                 |
| 16    | **Num_of_Delayed_Payment**   | Total instances of late payments made by the customer.                                          |
| 17    | **Changed_Credit_Limit**     | Adjustments made to the customer’s credit limit.                                               |
| 18    | **Num_Credit_Inquiries**     | Number of inquiries made regarding the customer's credit.                                       |
| 19    | **Credit_Mix**               | The variety of credit types the customer uses (e.g., loans, credit cards).                     |
| 20    | **Outstanding_Debt**         | Total amount of debt the customer currently owes.                                              |
| 21    | **Credit_Utilization_Ratio** | Proportion of credit used compared to the total credit limit.                                   |
| 22    | **Credit_History_Age**       | Duration of the customer’s credit history.                                                      |
| 23    | **Payment_of_Min_Amount**    | Indicates if the customer pays the minimum required amount each month.                         |
| 24    | **Total_EMI_per_month**      | Total Equated Monthly Installment (EMI) paid by the customer.                                   |
| 25    | **Amount_invested_monthly**  | Monthly investment amount made by the customer.                                                 |
| 26    | **Payment_Behaviour**        | Customer’s payment habits and tendencies.                                                       |
| 27    | **Monthly_Balance**          | The remaining balance in the customer’s account at the end of each month.                      |
| 28    | **Credit_Score**             | The customer’s credit score (target variable: "Good," "Poor," "Standard").                     |


- > *The original dataset is available on Kaggle:* [Credit Score Classification](https://www.kaggle.com/datasets/parisrohan/credit-score-classification/data?select=train.csv)

---
<p>


## Import Libraries

In [2]:
# Load the dataset
df_ = pd.read_csv("data/train.csv")  # Read the dataset from the specified path
df = df_.copy()  # Create a backup copy of the original dataset to preserve its integrity

C:\Users\USER\AppData\Local\Temp\ipykernel_17392\318680530.py:2: DtypeWarning: Columns (26) have mixed types. Specify dtype option on import or set low_memory=False.
  df_ = pd.read_csv("data/train.csv")  # Read the dataset from the specified path


## First look at the DATA

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 28 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   ID                        100000 non-null  object 
 1   Customer_ID               100000 non-null  object 
 2   Month                     100000 non-null  object 
 3   Name                      90015 non-null   object 
 4   Age                       100000 non-null  object 
 5   SSN                       100000 non-null  object 
 6   Occupation                100000 non-null  object 
 7   Annual_Income             100000 non-null  object 
 8   Monthly_Inhand_Salary     84998 non-null   float64
 9   Num_Bank_Accounts         100000 non-null  int64  
 10  Num_Credit_Card           100000 non-null  int64  
 11  Interest_Rate             100000 non-null  int64  
 12  Num_of_Loan               100000 non-null  object 
 13  Type_of_Loan              88592 non-null   ob

In [5]:
# Specify the columns to include in the analysis
columns_to_use = [
    "ID", "Customer_ID", "Month", "Name", "Age", "Credit_History_Age",
    "SSN", "Occupation", "Annual_Income", "Monthly_Inhand_Salary", "Total_EMI_per_month"
]

# selecting only the specified columns
df1 = df[columns_to_use]

df1.head(10)


,ID,Customer_ID,Month,Name,Age,Credit_History_Age,SSN,Occupation,Annual_Income,Monthly_Inhand_Salary,Total_EMI_per_month
0,0x1602,CUS_0xd40,January,Aaron Maashoh,23,22 Years and 1 Months,821-00-0265,Scientist,19114.12,1824.843333,49.574949
1,0x1603,CUS_0xd40,February,Aaron Maashoh,23,NaN,821-00-0265,Scientist,19114.12,NaN,49.574949
2,0x1604,CUS_0xd40,March,Aaron Maashoh,-500,22 Years and 3 Months,821-00-0265,Scientist,19114.12,NaN,49.574949
3,0x1605,CUS_0xd40,April,Aaron Maashoh,23,22 Years and 4 Months,821-00-0265,Scientist,19114.12,NaN,49.574949
4,0x1606,CUS_0xd40,May,Aaron Maashoh,23,22 Years and 5 Months,821-00-0265,Scientist,19114.12,1824.843333,49.574949
5,0x1607,CUS_0xd40,June,Aaron Maashoh,23,22 Years and 6 Months,821-00-0265,Scientist,19114.12,NaN,49.574949
6,0x1608,CUS_0xd40,July,Aaron Maashoh,23,22 Years and 7 Months,821-00-0265,Scientist,19114.12,1824.843333,49.574949
7,0x1609,CUS_0xd40,August,NaN,23,NaN,#F%$D@*&8,Scientist,19114.12,1824.843333,49.574949
8,0x160e,CUS_0x21b1,January,Rick Rothackerj,28_,26 Years and 7 Months,004-07-5839,_______,34847.84,3037.986667,18.816215
9,0x160f,CUS_0x21b1,February,Rick Rothackerj,28,26 Years and 8 Months,004-07-5839,Teacher,34847.84,3037.986667,18.816215


In [6]:
# Convert column names to lowercase for consistency and ease of use
df1.columns = [column.lower() for column in df1.columns]

In [7]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 11 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   id                     100000 non-null  object 
 1   customer_id            100000 non-null  object 
 2   month                  100000 non-null  object 
 3   name                   90015 non-null   object 
 4   age                    100000 non-null  object 
 5   credit_history_age     90970 non-null   object 
 6   ssn                    100000 non-null  object 
 7   occupation             100000 non-null  object 
 8   annual_income          100000 non-null  object 
 9   monthly_inhand_salary  84998 non-null   float64
 10  total_emi_per_month    100000 non-null  float64
dtypes: float64(2), object(9)
memory usage: 8.4+ MB


In [8]:
df1.describe(include="all").T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
id,100000,100000,0x1602,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
customer_id,100000,12500,CUS_0xd40,8,NaN,NaN,NaN,NaN,NaN,NaN,NaN
month,100000,8,January,12500,NaN,NaN,NaN,NaN,NaN,NaN,NaN
name,90015,10139,Langep,44,NaN,NaN,NaN,NaN,NaN,NaN,NaN
age,100000,1788,38,2833,NaN,NaN,NaN,NaN,NaN,NaN,NaN
credit_history_age,90970,404,15 Years and 11 Months,446,NaN,NaN,NaN,NaN,NaN,NaN,NaN
ssn,100000,12501,#F%$D@*&8,5572,NaN,NaN,NaN,NaN,NaN,NaN,NaN
occupation,100000,16,_______,7062,NaN,NaN,NaN,NaN,NaN,NaN,NaN
annual_income,100000,18940,36585.12,16,NaN,NaN,NaN,NaN,NaN,NaN,NaN
monthly_inhand_salary,84998.0,NaN,NaN,NaN,4194.17085,3183.686167,303.645417,1625.568229,3093.745,5957.448333,15204.633333


This dataset provides customer profiles and various financial details related to credit history. Upon examining the columns, several data issues stand out:

- **ID and Customer_ID**: Contain unique values, typically used for merging or grouping operations.
- **Month**: Useful for time-based customer behavior analysis; no adjustments required.
- **Name**: Some customers have missing names (`NaN`), which can be filled using other records associated with the same `Customer_ID`.
- **Age**: Contains outliers like negative values (e.g., -500), which need to be corrected.
- **SSN**: Some entries (e.g., w`#F%$D@*&8`) are invalid and which need to be corrected
- **Occupation**: Includes meaningless values (`_______`), which should be replaced with appropriate occupations.
- **Annual_Income**: Generally appears consistent but could benefit from outlier analysis for accuracy.
- **Monthly_Inhand_Salary**: Contains missing values (`NaN`), which can be estimated using annual income.
- **Credit_History_Age**: Missing values for some customers need to be imputed with appropriate methods.
- **Total_EMI_per_month**: Appears consistent but should be checked for extreme values.

Overall, the dataset contains missing values, outliers, and inconsistencies. Addressing these issues through cleaning and preprocessing will result in a more reliable dataset, crucial for accurate analysis and machine learning model development.

## ID

In [9]:
df1["id"].duplicated().sum()  # All id's are unique?

0

## CUSTOMER_ID

In [10]:
# Is the number of occurrences for each customer_id the same?
(df1["customer_id"].value_counts() == 8).all()

True

## MONTH

In [11]:
# Is the number of records (months) the same for each Customer id?
(df1.groupby("customer_id")["month"].count() == 8).all()

True

In [12]:
# Are all the month values equal to each other for each Customer id?
filtered_df = df1.groupby("customer_id")["month"].unique()

filtered_df.apply(lambda x: set(x) == set(filtered_df.iloc[0])).all()

True

In [13]:
df1["month"].value_counts()

month
January     12500
February    12500
March       12500
April       12500
May         12500
June        12500
July        12500
August      12500
Name: count, dtype: int64

## NAME

In [15]:
# Get the first 20 rows where the "age" column contains non-numeric values or missing data.
df1[pd.to_numeric(df1["age"], errors="coerce").isna()]["age"].head(20)

8      28_
54     34_
58     30_
71     24_
89     33_
97     35_
123    30_
184    31_
227    40_
236    37_
240    54_
256    21_
277    37_
311    37_
326    33_
359    20_
381    30_
387    43_
399    38_
524    18_
Name: age, dtype: object

In [16]:
df1['age'].unique()

array(['23', '-500', '28_', ..., '4808_', '2263', '1342'], dtype=object)

In [18]:
# Group by "customer_id" and count the occurrences of each "age" value within each group.
df1.groupby(["customer_id"]).age.value_counts().head(20)

customer_id  age 
CUS_0x1000   17      5
             18      3
CUS_0x1009   26      6
             25      2
CUS_0x100b   18      4
             19      4
CUS_0x1011   44      7
             43      1
CUS_0x1013   44      6
             43      2
CUS_0x1015   27      7
             4510    1
CUS_0x1018   15      8
CUS_0x1026   52      4
             51      4
CUS_0x102d   30      5
             31      3
CUS_0x102e   26      6
             25      2
CUS_0x1032   40      7
Name: count, dtype: int64

In [19]:
# Filter for rows where "age" is less than or equal to 100, then count the occurrences of each "age" value and sort them by age.
df1[df1['age'] <= 100]['age'].value_counts().sort_index()

age
14     1175
15     1574
16     1455
17     1502
18     2385
19     2793
20     2744
21     2716
22     2785
23     2654
24     2714
25     2861
26     2945
27     2859
28     2968
29     2735
30     2727
31     2955
32     2884
33     2543
34     2837
35     2866
36     2868
37     2742
38     2994
39     2846
40     2609
41     2785
42     2577
43     2734
44     2824
45     2642
46     1621
47     1227
48     1385
49     1375
50     1273
51     1291
52     1356
53     1354
54     1311
55     1366
56      362
95        3
99        1
100       1
Name: count, dtype: int64

- Under normal circumstances, human age typically has a **maximum of 100 years.** Therefore, we will consider only ages under 100 in the dataset.

In [30]:
df1.monthly_inhand_salary.info()

<class 'pandas.core.series.Series'>
RangeIndex: 100000 entries, 0 to 99999
Series name: monthly_inhand_salary
Non-Null Count  Dtype  
--------------  -----  
84998 non-null  float64
dtypes: float64(1)
memory usage: 781.4 KB


In [38]:
# Filter and display the first 20 invalid (non-numeric) values in the 'annual_income' column
df1[pd.to_numeric(df1["annual_income"], errors="coerce").isna()]["annual_income"].head(20)

10      34847.84_
27      30689.89_
32      35547.71_
56      34081.38_
66     114838.41_
94      88640.24_
98      54392.16_
106     8701.545_
111     8701.545_
115     25546.26_
129     92047.08_
151     97791.42_
175    10183.015_
180    106733.13_
188    12600.445_
197     57983.12_
214     34290.12_
217     43070.24_
231    6515990.0_
235     39641.54_
Name: annual_income, dtype: object

In [51]:
# Is there any other erroneous value?
filtered_df = df1.groupby(
    "customer_id")["total_emi_per_month"].nunique().loc[lambda x: x > 1].index

df1[df1["customer_id"].isin(filtered_df)].groupby(
    "customer_id")["total_emi_per_month"].value_counts().head(20)

Series([], Name: count, dtype: int64)

In [53]:
# Counting the occurrences of each value in the 'credit_history_age' column, including missing values (NaN),
# and displaying the top 20 most frequent values.
df1["credit_history_age"].value_counts(dropna=False).head(20)

credit_history_age
NaN                       9030
15 Years and 11 Months     446
19 Years and 4 Months      445
19 Years and 5 Months      444
17 Years and 11 Months     443
19 Years and 3 Months      441
17 Years and 9 Months      438
15 Years and 10 Months     436
17 Years and 10 Months     435
15 Years and 9 Months      432
18 Years and 3 Months      428
18 Years and 4 Months      426
18 Years and 2 Months      426
19 Years and 9 Months      422
17 Years and 8 Months      419
15 Years and 8 Months      415
18 Years and 11 Months     414
16 Years and 2 Months      412
18 Years and 5 Months      410
18 Years and 10 Months     408
Name: count, dtype: int64

In [56]:
df1["credit_history_age"].head(10)

0    265.0
1      NaN
2    267.0
3    268.0
4    269.0
5    270.0
6    271.0
7      NaN
8    319.0
9    320.0
Name: credit_history_age, dtype: float64

- In the function we wrote, we convert the years in the dataset into months, and we can observe the ordering of the values here. This helps in analyzing the data with a monthly perspective rather than yearly.

In [60]:
# Detecting invalid credit history data and marking it as NaN
def mark_invalid_credit_history(row):
    """
    Detects invalid credit history data and marks them as NaN:
    - The age should be greater than the credit history.
    - The credit history should be consistent with the age (the starting age should logically be 18).
    """
    age_in_months = row['age'] * 12
    if not (age_in_months - row['credit_history_age']) >= 18*12:
        return np.nan  # Invalid data is marked as NaN
    return row['credit_history_age']  # Valid data is kept

# Marking invalid data as NaN
df1['credit_history_age'] = df1.apply(mark_invalid_credit_history, axis=1)

# Checking the results
print("Updated column statistics:")
print(df1['credit_history_age'].describe())

Updated column statistics:
count    48737.000000
mean       174.132446
std         92.271479
min          1.000000
25%        100.000000
50%        168.000000
75%        234.000000
max        404.000000
Name: credit_history_age, dtype: float64


In [61]:
# Multiply age by 12 and observe alongside 'credit_history_age'
invalid_data_subset = invalid_data.copy()  # Create a copy of the invalid_data DataFrame
invalid_data_subset['age_in_months'] = invalid_data_subset['age'] * 12  # Convert age to months

# Observe the last 50 invalid records
invalid_data_subset = invalid_data_subset[['age_in_months', 'credit_history_age']].head(10)

print(invalid_data_subset)

   age_in_months  credit_history_age
0          276.0               265.0
1          276.0               266.0
2          276.0               267.0
3          276.0               268.0
4          276.0               269.0
5          276.0               270.0
6          276.0               271.0
7          276.0               272.0
8          336.0               319.0
9          336.0               320.0


In [65]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 100000 entries, 0 to 99999
Data columns (total 11 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   id                     100000 non-null  object 
 1   customer_id            100000 non-null  object 
 2   month                  100000 non-null  object 
 3   name                   100000 non-null  object 
 4   age                    100000 non-null  float64
 5   credit_history_age     100000 non-null  float64
 6   ssn                    100000 non-null  object 
 7   occupation             100000 non-null  object 
 8   annual_income          100000 non-null  float64
 9   monthly_inhand_salary  100000 non-null  float64
 10  total_emi_per_month    100000 non-null  float64
dtypes: float64(5), object(6)
memory usage: 11.2+ MB


- All the data in **Part 1** has been successfully filled without any data loss. As seen in the table, the number of rows in all columns should be 100,000, indicating that the data was properly handled and completed.

# PART 2

In [66]:
# Specify the columns to include in the analysis
columns_to_use = [
    "Customer_ID", "Num_Bank_Accounts", "Num_Credit_Card", "Interest_Rate",
    "Num_of_Loan", "Type_of_Loan", "Delay_from_due_date",
    "Num_of_Delayed_Payment", "Changed_Credit_Limit", "Num_Credit_Inquiries",
    "Credit_Mix", "Outstanding_Debt"
]
# Load the dataset, selecting only the specified columns
df2 = df[columns_to_use]

# Display the first five rows of the filtered dataset
df2.head(10)

,Customer_ID,Num_Bank_Accounts,Num_Credit_Card,Interest_Rate,Num_of_Loan,Type_of_Loan,Delay_from_due_date,Num_of_Delayed_Payment,Changed_Credit_Limit,Num_Credit_Inquiries,Credit_Mix,Outstanding_Debt
0,CUS_0xd40,3,4,3,4,"Auto Loan, Credit-Builder Loan, Personal Loan,...",3,7,11.27,4.0,_,809.98
1,CUS_0xd40,3,4,3,4,"Auto Loan, Credit-Builder Loan, Personal Loan,...",-1,NaN,11.27,4.0,Good,809.98
2,CUS_0xd40,3,4,3,4,"Auto Loan, Credit-Builder Loan, Personal Loan,...",3,7,_,4.0,Good,809.98
3,CUS_0xd40,3,4,3,4,"Auto Loan, Credit-Builder Loan, Personal Loan,...",5,4,6.27,4.0,Good,809.98
4,CUS_0xd40,3,4,3,4,"Auto Loan, Credit-Builder Loan, Personal Loan,...",6,NaN,11.27,4.0,Good,809.98
5,CUS_0xd40,3,4,3,4,"Auto Loan, Credit-Builder Loan, Personal Loan,...",8,4,9.27,4.0,Good,809.98
6,CUS_0xd40,3,4,3,4,"Auto Loan, Credit-Builder Loan, Personal Loan,...",3,8_,11.27,4.0,Good,809.98
7,CUS_0xd40,3,4,3,4,"Auto Loan, Credit-Builder Loan, Personal Loan,...",3,6,11.27,4.0,Good,809.98
8,CUS_0x21b1,2,4,6,1,Credit-Builder Loan,3,4,5.42,2.0,Good,605.03
9,CUS_0x21b1,2,4,6,1,Credit-Builder Loan,7,1,7.42,2.0,Good,605.03


Here is the general overview of the dataset summarized point by point
.
- **Num_Bank_Accounts**: The number of bank accounts per customer is mostly consistent. However, very high values (e.g., 20+ bank accounts) could be suspicious and should be checked for outliers during data cleaning.
- **Num_Credit_Card**: The number of credit cards is consistent, but similarly to bank accounts, unusually high numbers should be checked and cleaned if necessary.
- **Interest_Rate**: Interest rates should generally be numerical, but unusual (very low or very high) values should be checked. Outliers should be cleaned if necessary.
- **Num_of_Loan**: The number of loans taken by customers is consistent. However, if there are unusually large numbers of loans, it should be verified.
- **Type_of_Loan**: The loan type is categorical, indicating different types of loans. If there are inconsistencies or missing values, these should be addressed.
- **Delay_from_due_date**: The delay from the due date contains some negative values (-1), which are nonsensical and need to be corrected.
- **Num_of_Delayed_Payment**: The number of delayed payments contains missing values (NaN). These missing values should be filled using an appropriate method (e.g., median or mode).
- **Changed_Credit_Limit**: The changed credit limit is a numerical value. Missing or nonsensical values (e.g., `_` or extremely high/low limits) should be reviewed and corrected.
- **Num_Credit_Inquiries**: The number of credit inquiries is usually consistent, but unusually high numbers should be verified.
- **Credit_Mix**: The credit mix is typically categorical (e.g., "Good"). However, there are placeholders like `_` for missing or erroneous values, which should be addressed.
- **Outstanding_Debt**: Outstanding debt is numerical and generally consistent. Negative debt values or other nonsensical entries shou.inconsistent values.

In [67]:
# Convert column names to lowercase for consistency and ease of use
df2.columns = [column.lower() for column in df2.columns]

In [68]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 12 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   customer_id             100000 non-null  object 
 1   num_bank_accounts       100000 non-null  int64  
 2   num_credit_card         100000 non-null  int64  
 3   interest_rate           100000 non-null  int64  
 4   num_of_loan             100000 non-null  object 
 5   type_of_loan            88592 non-null   object 
 6   delay_from_due_date     100000 non-null  int64  
 7   num_of_delayed_payment  92998 non-null   object 
 8   changed_credit_limit    100000 non-null  object 
 9   num_credit_inquiries    98035 non-null   float64
 10  credit_mix              100000 non-null  object 
 11  outstanding_debt        100000 non-null  object 
dtypes: float64(1), int64(4), object(7)
memory usage: 9.2+ MB


## NUM_BANK_ACCOUNTS

In [69]:
# Group the DataFrame by 'customer_id'
# Calculate the value counts for combinations of 'num_bank_accounts' and 'num_credit_card' within each group
# Display the top 20 most frequent combinations across all customer groups
df2.groupby(["customer_id"])[["num_bank_accounts", "num_credit_card"]].value_counts().head(20)

customer_id  num_bank_accounts  num_credit_card
CUS_0x1000   6                  5                  8
CUS_0x1009   6                  5                  8
CUS_0x100b   1                  4                  8
CUS_0x1011   3                  3                  8
CUS_0x1013   3                  3                  8
CUS_0x1015   7                  4                  8
CUS_0x1018   7                  7                  8
CUS_0x1026   2                  6                  8
CUS_0x102d   5                  3                  7
             1000               3                  1
CUS_0x102e   8                  4                  8
CUS_0x1032   4                  7                  8
CUS_0x1037   5                  4                  8
CUS_0x1038   3                  4                  8
CUS_0x103e   4                  6                  8
CUS_0x1041   8                  8                  8
CUS_0x1044   2                  4                  8
CUS_0x1048   3                  7                  

- We are examining the consistency between credit card transactions and bank account balances to identify whether there are any outliers or significant discrepancies between them. This helps ensure data reliability and detect any extreme differences.

In [70]:
# Count unique values in 'num_bank_accounts', sort by value, and show the first 20.
df2["num_bank_accounts"].value_counts().sort_index().head(20)

num_bank_accounts
-1        21
 0      4328
 1      4490
 2      4304
 3     11950
 4     12186
 5     12118
 6     13001
 7     12823
 8     12765
 9      5443
 10     5247
 11        9
 18        1
 26        1
 27        1
 28        3
 29        1
 30        1
 31        1
Name: count, dtype: int64

- When we sort the data in ascending order, we observe that the value **-1** is incorrect and that calculations beyond **11** are nonsensical. Therefore, we will clean these erroneous values from the dataset.

In [73]:
# Count unique values in 'num_credit_card', sort by value, and show the first 20
df2["num_credit_card"].value_counts().sort_index().head(20)

num_credit_card
0        13
1      2132
2      2149
3     13277
4     14030
5     18459
6     16559
7     16615
8      4956
9      4643
10     4860
11       36
15        3
16        2
17        1
18        1
20        1
21        1
22        2
24        1
Name: count, dtype: int64

In [76]:
# Count unique values in 'interest_rat', sort by value, and show the first 50
df2.interest_rate.value_counts().sort_index().head(50)

interest_rate
1     2683
2     2465
3     2765
4     2589
5     4979
6     4721
7     4494
8     5012
9     4494
10    4540
11    4428
12    4540
13    2384
14    2229
15    3992
16    3730
17    3813
18    4102
19    3630
20    3929
21    1560
22    1720
23    1683
24    1685
25    1566
26    1489
27    1608
28    1616
29    1662
30    1690
31    1457
32    1742
33    1467
34    1502
59       1
60       1
64       1
67       1
73       1
74       1
75       2
76       2
77       1
81       1
87       1
89       1
91       1
92       1
94       1
95       2
Name: count, dtype: int64

In [80]:
# Convert 'num_of_loan' column to numeric, coercing errors to NaN
# Select rows where 'num_of_loan' could not be converted to a valid number (NaN values)
# Display the first 20 rows of the 'num_of_loan' column with NaN values
df2[pd.to_numeric(df2["num_of_loan"], errors="coerce").isna()]["num_of_loan"].head(20)

37     0_
80     3_
84     3_
94     2_
119    5_
125    2_
131    0_
144    0_
158    5_
256    2_
360    8_
366    8_
377    2_
382    2_
391    9_
410    5_
442    4_
481    7_
483    7_
486    7_
Name: num_of_loan, dtype: object

In [83]:
# Count unique values in 'num_of_loan', sort by value, and show the first 20
df2["num_of_loan"].value_counts().sort_index().head(20)

num_of_loan
-100     3876
 0      10930
 1      10606
 2      15032
 3      15104
 4      14743
 5       7197
 6       7803
 7       7344
 8       3191
 9       3702
 17         1
 18         1
 19         1
 23         2
 27         1
 29         1
 31         2
 32         1
 33         2
Name: count, dtype: int64

In [96]:
# Count the frequency of each unique value in 'delay_from_due_date' column
# Sort the counts by the unique values (index) and display the first 10
df2.delay_from_due_date.value_counts().sort_index().head(10)

delay_from_due_date
-5      33
-4      62
-3     118
-2     168
-1     210
 0    1195
 1    1326
 2    1342
 3    1686
 4    1722
Name: count, dtype: int64

In [97]:
# Group by 'customer_id', count the occurrences of each 'delay_from_due_date' value,
df2.groupby(["customer_id"])["delay_from_due_date"].value_counts().sample(20)

customer_id  delay_from_due_date
CUS_0x5a7b   61                     2
CUS_0xbb6c   65                     1
CUS_0x979c   15                     1
CUS_0x9a42   12                     1
CUS_0x5e06   4                      1
CUS_0x8588   16                     1
CUS_0x87ba   14                     1
CUS_0x26aa   53                     1
CUS_0x805b   17                     1
CUS_0xa59d   31                     1
CUS_0x4ebb   53                     4
CUS_0x4f26   23                     6
CUS_0x5d91   14                     6
CUS_0x27fb   3                      1
CUS_0x2cb1   17                     5
CUS_0x9ba5   16                     1
CUS_0x882f   62                     1
CUS_0x78cd   9                      1
CUS_0x3b3f   25                     7
CUS_0x2123   17                     1
Name: count, dtype: int64

- Based on our observations, we need to correct the **negative values** and **outliers** to ensure the data is accurate and consistent.

In [98]:
# Converts negative values in the 'delay_from_due_date' column to positive by applying the absolute value function
df2["delay_from_due_date"] = df2["delay_from_due_date"].abs()

C:\Users\USER\AppData\Local\Temp\ipykernel_17392\636789556.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2["delay_from_due_date"] = df2["delay_from_due_date"].abs()


In [99]:
# Counts the occurrences of each unique value in the 'delay_from_due_date' column, sorts them by index,
# and displays the first 10 entries
df2["delay_from_due_date"].value_counts().sort_index().head(10)

delay_from_due_date
0    1195
1    1536
2    1510
3    1804
4    1784
5    3075
6    3137
7    3234
8    3324
9    3233
Name: count, dtype: int64

In [100]:
# Verification of inaccurate data, checking if there is a difference larger than 30
# between consecutive values in the 'delay_from_due_date' column for each customer
def check_consecutive_delays(group):
    delays = group["delay_from_due_date"].values
    # Calculate the consecutive differences
    diff = abs(delays[1:] - delays[:-1])
    # Check if the absolute difference between consecutive delays is greater than 30
    return (diff > 30).any()


In [101]:
# Group by 'customer_id' and apply the function to check for large consecutive differences
df2.groupby("customer_id").apply(check_consecutive_delays) == True

customer_id
CUS_0x1000    False
CUS_0x1009    False
CUS_0x100b    False
CUS_0x1011    False
CUS_0x1013    False
              ...  
CUS_0xff3     False
CUS_0xff4     False
CUS_0xff6     False
CUS_0xffc     False
CUS_0xffd     False
Length: 12500, dtype: bool

- This script defines the check_consecutive_delays function that checks for significant discrepancies between consecutive values in the delay_from_due_date column for each customer. The function iterates over the delay values for each customer and calculates the absolute difference between consecutive delays. If the difference is greater than 30 days, it returns True, indicating a problem with the data. The script then groups the dataset by customer_id and applies this function to each group. Finally, it checks if any group (customer) has a delay difference greater than 30 and returns a boolean result and as you can see in the result, there are 12500 false positives, indicating that the function is working properly.

In [104]:
# Checking for rows where 'num_of_delayed_payment' cannot be converted to numeric
df2[pd.to_numeric(df2["num_of_delayed_payment"],
                  errors="coerce").isna()]["num_of_delayed_payment"].head(50)

1      NaN
4      NaN
6       8_
11      3_
30     NaN
32     NaN
33     NaN
53      3_
54      2_
70     NaN
95     NaN
106    NaN
119    13_
124    14_
140    NaN
141    12_
142    NaN
148    NaN
152    NaN
164    NaN
168    NaN
172    NaN
176     8_
194    NaN
201    NaN
202    NaN
209    NaN
269    NaN
271    NaN
274    NaN
298    22_
313    NaN
315    NaN
324    14_
328    NaN
336    NaN
353    NaN
356    NaN
358     4_
368    NaN
376    NaN
378    NaN
379    NaN
395    NaN
397    NaN
399    NaN
404    11_
411    NaN
418    NaN
420    21_
Name: num_of_delayed_payment, dtype: object

In [107]:
# Counting the frequency of each unique value in 'num_of_delayed_payment' and sorting by index
df2["num_of_delayed_payment"].value_counts().sort_index().head(50)

num_of_delayed_payment
-3.0        94
-2.0       234
-1.0       316
 0.0      1609
 1.0      1636
 2.0      1810
 3.0      1931
 4.0      1838
 5.0      2091
 6.0      2321
 7.0      2385
 8.0      4873
 9.0      4981
 10.0     5309
 11.0     4810
 12.0     5059
 13.0     4036
 14.0     4193
 15.0     5237
 16.0     5312
 17.0     5412
 18.0     5216
 19.0     5481
 20.0     5089
 21.0     2553
 22.0     2339
 23.0     2028
 24.0     1701
 25.0     1665
 26.0      322
 27.0      250
 28.0      131
 46.0        1
 47.0        1
 49.0        1
 52.0        1
 72.0        1
 80.0        1
 84.0        1
 85.0        1
 86.0        1
 88.0        1
 94.0        1
 98.0        1
 102.0       1
 106.0       1
 107.0       1
 121.0       1
 130.0       1
 132.0       1
Name: count, dtype: int64

In [113]:
# Get the value counts of 'changed_credit_limit', sorted by index, and display the first 20
df2["changed_credit_limit"].value_counts().sort_index().head(20)

changed_credit_limit
-6.49    1
-6.48    1
-6.45    1
-6.44    2
-6.43    2
-6.39    1
-6.37    1
-6.35    2
-6.33    1
-6.32    2
-6.31    4
-6.28    1
-6.27    1
-6.25    1
-6.23    1
-6.20    1
-6.19    2
-6.13    2
-6.10    1
-6.07    1
Name: count, dtype: int64

In [114]:
# Convert all values in 'changed_credit_limit' to their absolute values
df2["changed_credit_limit"] = df2["changed_credit_limit"].abs()

C:\Users\USER\AppData\Local\Temp\ipykernel_17392\3931560220.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2["changed_credit_limit"] = df2["changed_credit_limit"].abs()


## NUM_CREDIT_INQUIRIES

In [115]:
# Group by 'customer_id' and get unique values in the 'num_credit_inquiries' column for each customer
df2.groupby("customer_id")["num_credit_inquiries"].unique()

customer_id
CUS_0x1000          [10.0, 11.0]
CUS_0x1009            [2.0, 4.0]
CUS_0x100b    [1.0, 2271.0, 4.0]
CUS_0x1011         [7.0, 1965.0]
CUS_0x1013                 [3.0]
                     ...        
CUS_0xff3             [3.0, 5.0]
CUS_0xff4             [5.0, nan]
CUS_0xff6             [nan, 2.0]
CUS_0xffc            [8.0, 13.0]
CUS_0xffd     [1801.0, nan, 7.0]
Name: num_credit_inquiries, Length: 12500, dtype: object

In [116]:
# Get the value counts of 'num_credit_inquiries', including NaN values, and display the top 30
df2["num_credit_inquiries"].value_counts(dropna=False).head(30)

num_credit_inquiries
4.0       11271
3.0        8890
6.0        8111
7.0        8058
2.0        8028
8.0        7866
1.0        7588
0.0        6972
5.0        5693
9.0        5283
11.0       5051
10.0       4936
12.0       4571
NaN        1965
13.0       1480
14.0       1039
15.0        808
16.0        456
17.0        284
1786.0        4
318.0         4
1622.0        4
1114.0        4
1230.0        4
31.0          4
53.0          4
1768.0        4
1460.0        4
2297.0        4
285.0         4
Name: count, dtype: int64

In [120]:
# Group by customer_id and count unique values in credit_mix, then display the first 20
df2.groupby(["customer_id"])["credit_mix"].value_counts().head(20)

customer_id  credit_mix
CUS_0x1000   Bad           8
CUS_0x1009   Standard      5
             _             3
CUS_0x100b   Good          7
             _             1
CUS_0x1011   Standard      8
CUS_0x1013   Good          7
             _             1
CUS_0x1015   Standard      8
CUS_0x1018   Bad           6
             _             2
CUS_0x1026   Good          6
             _             2
CUS_0x102d   Good          7
             _             1
CUS_0x102e   Standard      6
             _             2
CUS_0x1032   Standard      8
CUS_0x1037   Good          7
             _             1
Name: count, dtype: int64

In [121]:
# Count how many customer_ids have exactly 3 unique credit_mix values
(df2.groupby(["customer_id"])["credit_mix"].nunique() == 3).sum()

0

In [126]:
# Identify rows where 'outstanding_debt' cannot be converted to numeric, and display the first 20
df2[pd.to_numeric(df2["outstanding_debt"], errors="coerce").isna()]["outstanding_debt"].head(20)

87      1328.93_
142     1283.37_
368     2797.17_
632     3818.57_
703      343.84_
731      363.51_
782      404.51_
819     1755.81_
835     2593.44_
992       89.62_
1147    1054.86_
1148    1054.86_
1162    1374.23_
1243    1166.47_
1339     457.81_
1369    1684.06_
1486    1126.35_
1542    2098.83_
1573     761.18_
1633    1343.01_
Name: outstanding_debt, dtype: object

In [128]:
# Count the unique values of 'outstanding_debt' and sort them in ascending order, then display the first 20 values
df2.outstanding_debt.value_counts().sort_index().head(20)

outstanding_debt
0.23     8
0.34     8
0.54     8
0.56     8
0.77     8
0.95    16
1.20     8
1.23     8
1.30     8
1.33     8
1.37     8
1.42     8
1.48     8
2.04     8
2.13     8
2.43     8
3.31     8
3.50     8
3.68     8
3.74     8
Name: count, dtype: int64

In [130]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 12 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   customer_id             100000 non-null  object 
 1   num_bank_accounts       100000 non-null  float64
 2   num_credit_card         100000 non-null  float64
 3   interest_rate           100000 non-null  float64
 4   num_of_loan             100000 non-null  float64
 5   type_of_loan            100000 non-null  object 
 6   delay_from_due_date     100000 non-null  int64  
 7   num_of_delayed_payment  100000 non-null  float64
 8   changed_credit_limit    100000 non-null  float32
 9   num_credit_inquiries    100000 non-null  float64
 10  credit_mix              100000 non-null  object 
 11  outstanding_debt        100000 non-null  float32
dtypes: float32(2), float64(6), int64(1), object(3)
memory usage: 8.4+ MB


- All the data in **Part 2** has been successfully filled without any data loss. As seen in the table, the number of rows in all columns should be 100,000, indicating that the data was properly handled and completed.

# PART 3

In [131]:
# Specify the columns to include in the analysis
columns_to_use = [
    "Customer_ID",
    "Credit_Utilization_Ratio",
    "Payment_of_Min_Amount",
    "Amount_invested_monthly",
    "Payment_Behaviour",
    "Monthly_Balance",
    "Credit_Score",
]
# Load the dataset, selecting only the specified columns
df3 = df[columns_to_use]
# Display the first five rows of the filtered datase
df3.head(10)

,Customer_ID,Credit_Utilization_Ratio,Payment_of_Min_Amount,Amount_invested_monthly,Payment_Behaviour,Monthly_Balance,Credit_Score
0,CUS_0xd40,26.822620,No,80.41529543900253,High_spent_Small_value_payments,312.49408867943663,Good
1,CUS_0xd40,31.944960,No,118.28022162236736,Low_spent_Large_value_payments,284.62916249607184,Good
2,CUS_0xd40,28.609352,No,81.699521264648,Low_spent_Medium_value_payments,331.2098628537912,Good
3,CUS_0xd40,31.377862,No,199.4580743910713,Low_spent_Small_value_payments,223.45130972736786,Good
4,CUS_0xd40,24.797347,No,41.420153086217326,High_spent_Medium_value_payments,341.48923103222177,Good
5,CUS_0xd40,27.262259,No,62.430172331195294,!@9#%8,340.4792117872438,Good
6,CUS_0xd40,22.537593,No,178.3440674122349,Low_spent_Small_value_payments,244.5653167062043,Good
7,CUS_0xd40,23.933795,No,24.785216509052056,High_spent_Medium_value_payments,358.12416760938714,Standard
8,CUS_0x21b1,24.464031,No,104.291825168246,Low_spent_Small_value_payments,470.69062692529184,Standard
9,CUS_0x21b1,38.550848,No,40.39123782853101,High_spent_Large_value_payments,484.5912142650067,Good


Here is an overview of the dataset with comments on each column:

- **Credit_Utilization_Ratio**: This column indicates the ratio of credit utilized by the customer relative to their available credit. The values range between 22.5% and 38.5%. These values appear consistent, with no missing data. However, it's important to monitor for outliers or extreme values, as they can indicate unusual spending patterns.

- **Payment_of_Min_Amount**: This categorical column reflects whether the customer has paid the minimum required amount or not. All entries show "No," which seems consistent. However, further analysis could be done to determine if a change in this behavior is necessary or if it is a systematic issue in the dataset.

- **Amount_invested_monthly**: This numerical column shows how much the customer invests monthly. The values vary from approximately 24.79 to 199.46, which seems normal, with no missing values. It's important to check for outliers or unrealistic values (e.g., very low or very high investments).

- **Payment_Behaviour**: This categorical column reflects the type of payment behavior of the customer. There are several categories like "High_spent_Small_value_payments," "Low_spent_Large_value_payments," and "Low_spent_Medium_value_payments." One value, "!@9#%8," seems like a placeholder or erroneous data and should be corrected or removed.

- **Monthly_Balance**: This numerical column represents the customer's monthly balance. The values are positive and vary between 223.45 and 484.59, which seem reasonable. It's important to ensure no negative or unusually large values are present.

- **Credit_Score**: This categorical column indicates the customer's credit score, with two values: "Good" and "Standard." The credit score values seem consistent, though there might be other categories in the dataset not seen here, which should be checked for completeness.

In [132]:
# Convert column names to lowercase for consistency and ease of use
df3.columns = [column.lower() for column in df3.columns]

In [133]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 7 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   customer_id               100000 non-null  object 
 1   credit_utilization_ratio  100000 non-null  float64
 2   payment_of_min_amount     100000 non-null  object 
 3   amount_invested_monthly   95521 non-null   object 
 4   payment_behaviour         100000 non-null  object 
 5   monthly_balance           98800 non-null   object 
 6   credit_score              100000 non-null  object 
dtypes: float64(1), object(6)
memory usage: 5.3+ MB


## CREDIT_UTILIZATION_RATIO

In [149]:
# Step 1: Get customer_ids who have invested exactly 10000 at least once
customer_with_10000 = df3.loc[df3["amount_invested_monthly"] == 10000, "customer_id"].unique()

# Step 2: Filter the entire dataset to only include those customers
df_filtered = df3[df3["customer_id"].isin(customer_with_10000)]

# Step 3: Group by 'customer_id' and 'amount_invested_monthly', and count occurrences
result = (
    df_filtered.groupby(["customer_id", "amount_invested_monthly"])
    .size()
    .to_frame(name="count")
)

print(result.head(50))

                                     count
customer_id amount_invested_monthly       
CUS_0x100b  174.287598                   1
            397.241852                   1
            424.003204                   1
            617.079285                   1
            661.624023                   1
            758.083191                   1
            10000.000000                 1
CUS_0x1011  71.226166                    1
            74.567383                    1
            119.853088                   1
            383.350830                   1
            444.197815                   1
            523.234009                   1
            560.997070                   1
            10000.000000                 1
CUS_0x1041  12.035500                    1
            19.626997                    1
            29.743185                    1
            37.888050                    1
            42.757725                    1
            43.565857                    1
           

In [150]:
# Check if all values in 'amount_invested_monthly' are NaN for each 'customer_id'
df3.groupby("customer_id")["amount_invested_monthly"].apply(
    lambda x: x.isna().all()).sum()

0

- In all **customer_id** groups, no group in the **amount_invested_monthly** column consists entirely of missing (NaN) values. This ensures that there is some data available for every group, and no group is entirely incomplete in this specific column.

In [154]:
# Group the DataFrame by 'customer_id', then calculate the number of unique values in the 'payment_behaviour' column for each customer
df3.groupby(["customer_id"])["payment_behaviour"].nunique().head(10)

customer_id
CUS_0x1000    5
CUS_0x1009    4
CUS_0x100b    5
CUS_0x1011    4
CUS_0x1013    4
CUS_0x1015    5
CUS_0x1018    4
CUS_0x1026    5
CUS_0x102d    4
CUS_0x102e    5
Name: payment_behaviour, dtype: int64

In [155]:
# Group the DataFrame by 'customer_id', calculate the number of unique values in 'payment_behaviour' for each customer,
# and then count how many customers have the same number of unique payment behaviors.
df3.groupby(["customer_id"])["payment_behaviour"].nunique().value_counts()

payment_behaviour
5    4448
4    4426
3    1632
6    1616
2     220
7     148
1      10
Name: count, dtype: int64

**Most Customers Have 4 or 5 Unique Behaviors:**

Most customers (4448 and 4426 customers) have 4 or 5 different **payment_behaviour** values. This indicates that these customers exhibit a variety of payment behaviors.

**Customers with Fewer Unique Behaviors:**

Customers with 3, 6, 2, 7, and 1 unique values are less common. This suggests that these customers may have more limited or consistent payment behaviors.

**Customers with Only 1 Unique Behavior:**

The 10 customers with only 1 unique behavior may demonstrate very little diversity in their payment patterns, possibly indicating that they exhibit just one payment behavior.

**Differences in Customer Payment Behaviors:**

The high number of customers with 5 and 4 unique behaviors suggests that, generally, there is diversity in payment behaviors. This could indicate that customers show different payment habits at different times or experience changes in their payment behavior.

**Customers with Limited Unique Behaviors:**

With only 1 unique **payment_behaviour** value, the very few (10 people) may have very stable payment habits, showing little to no variation.

This analysis is helpful for understanding the diversity in customer payment habits and could be important for assessing financial risk or managing customer relationships.

In [171]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 7 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   customer_id               100000 non-null  object 
 1   credit_utilization_ratio  100000 non-null  float64
 2   payment_of_min_amount     100000 non-null  object 
 3   amount_invested_monthly   100000 non-null  float32
 4   payment_behaviour         100000 non-null  object 
 5   monthly_balance           100000 non-null  float64
 6   credit_score              100000 non-null  int32  
dtypes: float32(1), float64(2), int32(1), object(3)
memory usage: 4.6+ MB


- All the data in **Part 3** has been successfully filled without any data loss. As seen in the table, the number of rows in all columns should be 100,000, indicating that the data was properly handled and completed.

# DATA VISUALIZATION

In [172]:
# Join df1 with df2 (after removing the 'customer_id' column from df2) and then join with df3 (after removing 'customer_id' from df3).
df = df1.join(df2.drop(columns=['customer_id'])).join(df3.drop(columns=['customer_id']))

# Remove duplicate columns, if any, keeping only the first occurrence of each column.
df = df.loc[:, ~df.columns.duplicated()]

# Display the first few rows of the final DataFrame.
df.head()

,id,customer_id,month,name,age,credit_history_age,ssn,occupation,annual_income,monthly_inhand_salary,...,changed_credit_limit,num_credit_inquiries,credit_mix,outstanding_debt,credit_utilization_ratio,payment_of_min_amount,amount_invested_monthly,payment_behaviour,monthly_balance,credit_score
0,0x1602,CUS_0xd40,January,Aaron Maashoh,23.00,29.00,821-00-0265,Scientist,19114.12,1824.84,...,11.27,4.00,Good,809.98,26.82,No,80.42,High_spent_Small_value_payments,312.49,2
1,0x1603,CUS_0xd40,February,Aaron Maashoh,23.00,14.00,821-00-0265,Scientist,19114.12,1824.84,...,11.27,4.00,Good,809.98,31.94,No,118.28,Low_spent_Large_value_payments,284.63,2
2,0x1604,CUS_0xd40,March,Aaron Maashoh,23.00,3.00,821-00-0265,Scientist,19114.12,1824.84,...,11.27,4.00,Good,809.98,28.61,No,81.70,Low_spent_Medium_value_payments,331.21,2
3,0x1605,CUS_0xd40,April,Aaron Maashoh,23.00,11.00,821-00-0265,Scientist,19114.12,1824.84,...,6.27,4.00,Good,809.98,31.38,No,199.46,Low_spent_Small_value_payments,223.45,2
4,0x1606,CUS_0xd40,May,Aaron Maashoh,23.00,2.00,821-00-0265,Scientist,19114.12,1824.84,...,11.27,4.00,Good,809.98,24.80,No,41.42,High_spent_Medium_value_payments,341.49,2


In [173]:
def optimize_dtypes(df):
    # Iterate over numeric columns
    for col in df.select_dtypes(include=[np.number]).columns:
        if pd.api.types.is_float_dtype(df[col]):
            if (df[col] % 1 == 0).all():  # Check if all float values are whole numbers
                # Convert float to the smallest suitable integer type
                df[col] = pd.to_numeric(df[col], downcast='integer')
            else:
                # Convert to the smallest suitable float type
                df[col] = pd.to_numeric(df[col], downcast='float')
        elif pd.api.types.is_integer_dtype(df[col]):
            # Convert integer to the smallest suitable integer type
            df[col] = pd.to_numeric(df[col], downcast='integer')

    return df

# Example usage:
df = optimize_dtypes(df)

- This code was written to optimize the data types of numeric columns in a pandas DataFrame. Its purpose is to reduce memory usage and improve performance by shrinking unnecessarily large data types. For example, it converts columns of type `int64` or `float64` to smaller and more appropriate data types such as `int8`, `int16`, `int32` or `float32` depending on the data range This saves memory and speeds up the process.

In [174]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 100000 entries, 0 to 99999
Data columns (total 28 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   id                        100000 non-null  object 
 1   customer_id               100000 non-null  object 
 2   month                     100000 non-null  object 
 3   name                      100000 non-null  object 
 4   age                       100000 non-null  int8   
 5   credit_history_age        100000 non-null  int16  
 6   ssn                       100000 non-null  object 
 7   occupation                100000 non-null  object 
 8   annual_income             100000 non-null  float64
 9   monthly_inhand_salary     100000 non-null  float32
 10  total_emi_per_month       100000 non-null  float32
 11  num_bank_accounts         100000 non-null  int8   
 12  num_credit_card           100000 non-null  int8   
 13  interest_rate             100000 non-null  int8   

In [175]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
age,100000.00,35.20,9.16,20.00,28.00,35.00,42.00,56.00
credit_history_age,100000.00,122.97,93.47,0.00,44.00,104.00,188.00,404.00
annual_income,100000.00,50505.12,38299.42,7005.93,19342.97,36999.71,71683.47,179987.28
monthly_inhand_salary,100000.00,4198.77,3187.49,303.65,1626.76,3096.38,5961.75,15204.63
total_emi_per_month,100000.00,105.54,125.81,0.00,29.05,66.03,145.58,1779.10
num_bank_accounts,100000.00,5.37,2.59,0.00,3.00,5.00,7.00,11.00
num_credit_card,100000.00,5.53,2.07,0.00,4.00,5.00,7.00,10.00
interest_rate,100000.00,14.53,8.74,1.00,7.00,13.00,20.00,34.00
num_of_loan,100000.00,3.53,2.45,0.00,2.00,3.00,5.00,9.00
delay_from_due_date,100000.00,21.10,14.82,0.00,10.00,18.00,28.00,67.00


In [182]:
from sklearn.preprocessing import MultiLabelBinarizer

df4["type_of_loan"] = df4["type_of_loan"].apply(lambda x: x if isinstance(x, list) else [])
mlb = MultiLabelBinarizer()
loan_encoded = mlb.fit_transform(df4["type_of_loan"])
loan_encoded_df = pd.DataFrame(loan_encoded, columns=mlb.classes_)
loan_encoded_df.head()
df_cat = loan_encoded_df
# Ghép lại với DataFrame gốc
# df_cat = pd.concat([df4.drop(columns=["type_of_loan"]), loan_encoded_df], axis=1)


In [189]:
df = pd.concat([df_cat.reset_index(drop=True), df.reset_index(drop=True).drop(['credit_mix', 'payment_of_min_amount', 'payment_behaviour'], axis = 1)], axis=1)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 38 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   Auto Loan                 100000 non-null  int32  
 1   Credit-Builder Loan       100000 non-null  int32  
 2   Debt Consolidation Loan   100000 non-null  int32  
 3   Home Equity Loan          100000 non-null  int32  
 4   Mortgage Loan             100000 non-null  int32  
 5   No Loan                   100000 non-null  int32  
 6   Not Specified             100000 non-null  int32  
 7   Payday Loan               100000 non-null  int32  
 8   Personal Loan             100000 non-null  int32  
 9   Student Loan              100000 non-null  int32  
 10  credit_mix                100000 non-null  int64  
 11  payment_of_min_amount     100000 non-null  int64  
 12  payment_behaviour         100000 non-null  int64  
 13  id                        100000 non-null  ob

In [190]:
df = df.drop([
    "id", "customer_id", "name", "ssn", "type_of_loan", 'month', 'occupation'], axis=1)

df.head()

,Auto Loan,Credit-Builder Loan,Debt Consolidation Loan,Home Equity Loan,Mortgage Loan,No Loan,Not Specified,Payday Loan,Personal Loan,Student Loan,...,num_of_loan,delay_from_due_date,num_of_delayed_payment,changed_credit_limit,num_credit_inquiries,outstanding_debt,credit_utilization_ratio,amount_invested_monthly,monthly_balance,credit_score
0,1,1,0,1,0,0,0,0,1,0,...,4,3,7,11.27,4,809.98,26.82,80.42,312.49,2
1,1,1,0,1,0,0,0,0,1,0,...,4,1,7,11.27,4,809.98,31.94,118.28,284.63,2
2,1,1,0,1,0,0,0,0,1,0,...,4,3,7,11.27,4,809.98,28.61,81.70,331.21,2
3,1,1,0,1,0,0,0,0,1,0,...,4,5,4,6.27,4,809.98,31.38,199.46,223.45,2
4,1,1,0,1,0,0,0,0,1,0,...,4,6,4,11.27,4,809.98,24.80,41.42,341.49,2


In [284]:
y_test.head()

29322    1
23872    1
59885    0
54432    1
47864    0
Name: credit_score, dtype: int8

# Decision Tree